beamer/tests/agent/multiple_processors/test_multiple_event_processors.py

```diff
@@ -0,0 +1,241 @@
+        rpc_urls[str(chain_id)] = rpc_url
+
+    token_list = _get_token_list(chain_map, allowance, slave_contract_addresses)
+    config = Config(
+        rpc_urls=rpc_urls,
+        deployment_info=deployment_info,
+        token_checker=TokenChecker(token_list),
+        account=account,
+        fill_wait_time=0,
+        unsafe_fill_time=600,
+        prometheus_metrics_port=None,
+        log_level="debug",
+    )
+    beamer.agent.metrics.init(config=config, source_rpc_url=url, target_rpc_url=url)
+    return config
+
+
+def _start_agent_test(config):
+    agent = Agent(config)
+    agent.start()
+
+    directions = agent.get_directions()
+
+    for direction in directions:
+        with Sleeper(5) as sleeper:
+            while len(agent.get_context(direction).requests) != 1:
+                sleeper.sleep(0.1)
+
+            assert len(agent.get_context(direction).requests) == 1
+
+            request = list(agent.get_context(direction).requests)[0]
+
+            with Sleeper(5) as sleeper:
+                while not request.is_claimed:
+                    sleeper.sleep(0.1)
+    agent.stop()
+
+
+def _get_slave_contract_addresses(proc: psutil.Popen):
+    while True:
+        out = proc.stdout.readline().decode()
+        print(out, flush=True)
+        if "Chain is ready" in out:
+            proc.stdin.write("get_contracts\n".encode())
+            proc.stdin.flush()
+            out = proc.stdout.readline().decode()
+            contract_addresses = json.loads(out.replace("\n", ""))
+            return contract_addresses
+
+
+def _stop_slave_tests(slave_procs: list[psutil.Popen]):
+    for proc in slave_procs:
+        os.kill(proc.pid, signal.SIGKILL)
+        proc.wait()
+
+
+def _get_token_list(chain_map, allowance, slave_contract_addresses):
+    tokens = []
+    for chain_id in chain_map:
+        if allowance is None:
+            tokens.append([chain_id, slave_contract_addresses[chain_id]["token"]])
+        else:
+            tokens.append([chain_id, slave_contract_addresses[chain_id]["token"], allowance])
+    return [tokens]
+
+
+def _mint_agent_tokens(config, token, slave_contract_addresses):
+    for chain_name, chain_rpc in config.rpc_urls.items():
+        if chain_name == "l1":
+            continue
+        w3 = make_web3(chain_rpc, config.account)
+        while not w3.is_connected():
+            time.sleep(1)
+        l2_token = w3.eth.contract(
+            address=slave_contract_addresses[int(chain_name)]["token"],
+            abi=[abi.dict() for abi in token.contract_type.abi],
+        )
+        transact(l2_token.functions.mint(config.account.address, 300))
+
+
+def test_multiple_event_processor(contracts, token, allowance):
+    chain_map = _get_chain_map()
+    with _new_networks(chain_map) as ape_config:
+        slave_test_procs = []
+        slave_contract_addresses = {}
+        for chain_id, port in chain_map.items():
+            slave = _start_slave_test(port, len(chain_map.values()))
```

Comment:
`len(chain_map.values())` is the same as `len(chain_map)` (only less efficient).